source/common/upstream/health_checker_impl.cc

```diff
@@ -20,47 +20,47 @@
 #include "common/http/header_map_impl.h"
 #include "common/http/headers.h"
 #include "common/http/utility.h"
-#include "common/json/config_schemas.h"
-#include "common/json/json_loader.h"
+#include "common/protobuf/utility.h"
 #include "common/redis/conn_pool_impl.h"
 #include "common/upstream/host_utility.h"
 
 namespace Envoy {
 namespace Upstream {
 
-HealthCheckerPtr HealthCheckerFactory::create(const Json::Object& hc_config,
+HealthCheckerPtr HealthCheckerFactory::create(const envoy::api::v2::HealthCheck& hc_config,
                                               Upstream::Cluster& cluster, Runtime::Loader& runtime,
                                               Runtime::RandomGenerator& random,
                                               Event::Dispatcher& dispatcher) {
-  hc_config.validateSchema(Json::Schema::CLUSTER_HEALTH_CHECK_SCHEMA);
-
-  const std::string hc_type = hc_config.getString("type");
-  if (hc_type == "http") {
+  switch (hc_config.health_checker_case()) {
+  case envoy::api::v2::HealthCheck::HealthCheckerCase::kHttpHealthCheck:
     return HealthCheckerPtr{
         new ProdHttpHealthCheckerImpl(cluster, hc_config, dispatcher, runtime, random)};
-  } else if (hc_type == "tcp") {
+  case envoy::api::v2::HealthCheck::HealthCheckerCase::kTcpHealthCheck:
     return HealthCheckerPtr{
         new TcpHealthCheckerImpl(cluster, hc_config, dispatcher, runtime, random)};
-  } else {
-    ASSERT(hc_type == "redis");
+  case envoy::api::v2::HealthCheck::HealthCheckerCase::kRedisHealthCheck:
     return HealthCheckerPtr{
         new RedisHealthCheckerImpl(cluster, hc_config, dispatcher, runtime, random,
                                    Redis::ConnPool::ClientFactoryImpl::instance_)};
+  default:
+    throw EnvoyException("Health checker not set");
```

Comment:
Yep, will change the wording and add a TODO, this is where the constraint checks would help. I would even be comfortable with just an ASSERT today, since this should already be picked up by the JSON schema on the JSON-Proto path. I will have a PR ready soon to turn on CDS/EDS for v2 via gRPC as an extreme experimental feature without the constraint checks, so we will need to figure this out sooner rather than later.